samples/03.speech/h.select-audio-input-device/comprehensive.html

```diff
@@ -0,0 +1,246 @@
+        // and to understand the risks associated with using secrets, visit https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication?view=azure-bot-service-4.0
+
+        const res = await fetch('https://webchat-mockbot.azurewebsites.net/directline/token', { method: 'POST' });
+        const { token } = await res.json();
+
+        // Imports
+        const {
+          navigator: { mediaDevices },
+          React: { useCallback, useEffect, useMemo, useState },
+          ReactDOM,
+          WebChat: { createCognitiveServicesSpeechServicesPonyfillFactory, createDirectLine, ReactWebChat }
+        } = window;
+
+        const App = () => {
+          // List of audio input devices. The list is initially empty until the async function call finished.
+          // If the browser does not support media devices, we will set it to false to show an error message.
+          const [audioInputDevices, setAudioInputDevices] = useState(mediaDevices ? [] : false);
+
+          // Time when the device list change. This is used to trigger a re-enumeration of devices.
+          const [lastDeviceChangeAt, setLastDeviceChangeAt] = useState(0);
+
+          // Device ID of the audio input device selected by the user.
+          const [selectedAudioInputDeviceId, setSelectedAudioInputDeviceId] = useState('default');
+
+          // Direct Line adapter instance will be kept across render.
+          const directLine = useMemo(() => createDirectLine({ token }), [token]);
+
+          // Function instance to fetch credentials for Cognitive Services Speech Services. The instance is kept across render to cache the credentials.
+          const fetchSpeechServicesCredentials = useMemo(() => createFetchSpeechServicesCredentials(), []);
+
+          // Create the ponyfill factory function, which can be called to create a concrete implementation of the ponyfill.
+          const webSpeechPonyfillFactory = useMemo(() =>
+            createCognitiveServicesSpeechServicesPonyfillFactory(
+              {
+                // Device ID of the audio input device to use.
+                audioInputDeviceId: selectedAudioInputDeviceId,
+
+                // We are passing the Promise function to the "credentials" field.
+                // This function will be called every time the token is being used.
+                credentials: fetchSpeechServicesCredentials
+              },
+              [selectedAudioInputDeviceId, fetchSpeechServicesCredentials]
+            )
+          );
+
+          // When the user select a different device, we change the "selectedAudioInputDeviceId" based on their selection.
+          const handleSelectedAudioInputDeviceIdChange = useCallback(({ target: { value } }) =>
+            setSelectedAudioInputDeviceId(value)
+          );
+
+          // When device is plugged/unplugged, we will re-enumerate the device list.
```

Comment:
```suggestion
          // When device is connected/disconnected, we will re-enumerate the device list.
```

Reply:
For wired devices, yes. But this could include Bluetooth so that's why I prefer connected/disconnected. I see what you mean though, so I'll resolve.